File: backend/modelarium-user-service/src/main/java/org/modelarium/user/model/UserEntity.java
package org.modelarium.user.model;

import jakarta.persistence.*;
import lombok.*;

import java.time.OffsetDateTime;
import java.util.UUID;

@Entity
@Table(name = "users")
@Getter
@Setter
@Builder(builderMethodName = "builder")
@NoArgsConstructor(access = AccessLevel.PROTECTED)
@AllArgsConstructor
public class UserEntity {
    @Id
    @Column(columnDefinition = "uuid", updatable = false, nullable = false)
    private UUID id;
    @Column(name = "sequence_id", updatable = false, nullable = false, unique = true)
    private Long sequenceId;
    @Column(name = "username", unique = true, nullable = false)
    private String userName;
    @Column(unique = true, nullable = false)
    private String email;
    @Column(name = "password_hash", nullable = false)
    private String passwordHash;
    @Column(name = "display_name")
    private String displayName;
    @Column(name = "avatar_key")
    private String avatarKey;
    @Column(name = "bio")
    private String bio;
    @Column(name = "email_verified", nullable = false)
    private boolean emailVerified;
    @Column(name = "created_at", nullable = false, updatable = false)
    private OffsetDateTime createdAt;
    @Column(name = "updated_at", nullable = false, updatable = false)
    private OffsetDateTime updateAt;
    @Column(name = "popularity_index", nullable = false, updatable = true)
    private int popularityIndex;

    @PrePersist
    protected void onCreate() {
        if (id == null) {
            id = UUID.randomUUID();
            popularityIndex = 0;
        }
        createdAt = OffsetDateTime.now();
        updateAt = createdAt;
    }

    @PreUpdate
    protected void onUpdate() {
        updateAt = OffsetDateTime.now();
    }
}

File: backend/modelarium-auth-service/src/main/kotlin/persistence/entity/UserEntity.kt
package org.modelarium.auth.persistence.entity

import jakarta.persistence.Column
import jakarta.persistence.Entity
import jakarta.persistence.Id
import jakarta.persistence.Table
import java.time.OffsetDateTime
import java.util.UUID

@Entity
@Table(name = "user", schema = "public")
class UserEntity(
    @Id
    @Column(columnDefinition = "uuid")
    val id: UUID = UUID.randomUUID(),
    @Column(name = "username", unique = true, nullable = false)
    var userName: String,
    @Column(unique = true, nullable = false)
    var email: String,
    @Column(name = "password_hash", nullable = false)
    var passwordHash: String,
    @Column(name = "display_name")
    var displayName: String? = null,
    @Column(name = "avatar_key")
    var avatarKey: String? = null,
    @Column(name = "bio")
    var bio: String? = null,
    @Column(name = "email_verified")
    var emailVerified: Boolean = false,
    @Column(name = "created_at")
    var createdAt: OffsetDateTime = OffsetDateTime.now(),
    @Column(name = "updated_at")
    var updatedAt: OffsetDateTime = OffsetDateTime.now()
)